src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,352 @@
+                => _codeStyleProvider.FixAllAsync(document, diagnostics, editor, cancellationToken);
+        }
+
+        private async Task RegisterCodeFixesAsync(
+            CodeStyle2CodeFixProvider fixProvider,
+            CodeFixContext context)
+        {
+            _fixProvider = fixProvider;
+
+            var document = context.Document;
+            var diagnostic = context.Diagnostics[0];
+            var cancellationToken = context.CancellationToken;
+
+            var codeFixes = await ComputeCodeActionsAsync(
+                document, diagnostic, cancellationToken).ConfigureAwait(false);
+            context.RegisterFixes(codeFixes, context.Diagnostics);
+        }
+
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeCodeActionsAsync(
+            Document document, Diagnostic diagnostic, CancellationToken cancellationToken);
+
+        protected abstract Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        protected Task<Document> FixAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+            => _fixProvider.FixAsync(document, diagnostic, cancellationToken);
+
+        #endregion
+    }
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class ExampleDiagnosticAnalyzer : ExampleCodeStyleProvider.CodeStyle2DiagnosticAnalyzer
+    {
+    }
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = "Example code refactoring provider")]
+    class ExampleCodeRefactoringProvider : ExampleCodeStyleProvider.CodeStyle2CodeRefactoringProvider
+    {
+    }
+
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class ExampleCodeFixProvider : ExampleCodeStyleProvider.CodeStyle2CodeFixProvider
+    {
+    }
+
+    class ExampleCodeStyleProvider : AbstractCodeStyle2Provider<ExpressionBodyPreference, int, ExampleCodeStyleProvider>
+    {
+        public ExampleCodeStyleProvider()
+            : base(SomeOption,
+                   IDEDiagnosticIds.SomeID,
+                   SomeTitle,
+                   SomeMessage)
+        {
+        }
+
+        #region analysis
+
+        protected override bool DiagnosticsForOpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override (ImmutableArray<int>, Action<SyntaxNodeAnalysisContext, ExpressionBodyPreference>) GetSyntaxNodeAction()
+        {
+            return (ImmutableArray.Create(1, 2, 3), AnalyzeSyntax);
+        }
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, ExpressionBodyPreference preference)
+        {
+            throw new NotImplementedException();
+        }
```

Comment:
subclass would implement this. but it would only actually be called if appropriate. subclass would not have to actually have to have the logic about "should i be runnign... or should only teh refactoring be running?" :)